File: weightpicker/src/main/java/co/gbyte/android/weightpicker/WeightPicker.java
package co.gbyte.android.weightpicker;

import android.app.Activity;
import android.content.Context;
import android.content.res.TypedArray;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.KeyEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.NumberPicker;
import android.widget.RelativeLayout;
import android.widget.TextView;

import java.text.DecimalFormat;
import java.text.NumberFormat;
import java.util.Locale;

/**
 * Created by walt on 23/01/17.
 *
 */

public class WeightPicker extends RelativeLayout {
    private static final String SUPERSTATE = "superState";
    private static final String VALUE = "value";

    private final int MIN_MIN_VALUE = 0;
    private final int MAX_MAX_VALUE = 999949;

    private int mPrecision = 1;
    private int mBase = 1;
    private int mMinValue = MIN_MIN_VALUE;
    private int mMaxValue = MAX_MAX_VALUE;
    private int mOldValue = mMinValue;

    private NumberPicker mKilograms = null;
    private NumberPicker mHectograms = null;
    private EditText mPickerEdit = null;
    private boolean mInitialSetup = true;

    public OnValueChangedListener mListener;

    public WeightPicker(Context context) {
        this(context, null);
    }
    public WeightPicker(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }
    public WeightPicker(final Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);

        final Activity activity = (Activity) getContext();
        activity.getLayoutInflater().inflate(R.layout.weightpicker, this, true);

        TextView separator = (TextView) findViewById(R.id.separator);
        NumberFormat numberFormat = NumberFormat.getInstance(Locale.getDefault());
        if (numberFormat instanceof DecimalFormat) {
            char sep = ((DecimalFormat) numberFormat).getDecimalFormatSymbols().getDecimalSeparator();
            separator.setText(String.format("%c", sep));
        }

        mHectograms = (NumberPicker) findViewById(R.id.hectograms);
        mKilograms  = (NumberPicker) findViewById(R.id.kilograms);
        mPickerEdit = (EditText) findViewById(R.id.weightPickerEdit);

        mPickerEdit.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView textView, int i, KeyEvent keyEvent) {

                if(i == EditorInfo.IME_ACTION_DONE) {
                    mPickerEdit.setVisibility(GONE);
                }
                mPickerEdit.clearFocus();
                return true;
            }
        });

        mPickerEdit.setSelectAllOnFocus(true);
        mPickerEdit.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if (hasFocus) {
                    mOldValue = WeightPicker.this.getValue();
                    ((InputMethodManager) activity.getSystemService(Context.INPUT_METHOD_SERVICE))
                            .toggleSoftInput(InputMethodManager.SHOW_FORCED,0);
                    mPickerEdit.selectAll();
                } else {
                    activity.getWindow().setSoftInputMode(WindowManager
                            .LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN);
                    double value = Double.parseDouble(mPickerEdit.getText().toString());
                    setValue((int) (value * mPrecision * mBase));

                    mListener.onValueChange(WeightPicker.this, mOldValue, getValue());
                    mPickerEdit.setVisibility(GONE);
                }
            }
        });

        OnLongClickListener longClickListener = new OnLongClickListener() {
            @Override
            public boolean onLongClick(View view) {
                mPickerEdit.setVisibility(VISIBLE);
                mPickerEdit.setText(String.format(Locale.getDefault(),
                                    "%.1f",
                                    (double) getValue() / (mPrecision * mBase)));
                mPickerEdit.requestFocus();

                return true;
            }
        };

        mHectograms.setOnLongClickListener(longClickListener);
        separator.setOnLongClickListener(longClickListener);
        mKilograms.setOnLongClickListener(longClickListener);

        EditText hectogramsEditText = findInput(mHectograms);
        if (hectogramsEditText != null) {
            hectogramsEditText.setFocusable(false);
        }

        EditText kilogramsEditText = findInput(mKilograms);
        if (kilogramsEditText != null) {
            kilogramsEditText.setFocusable(false);
        }

        // Find out the background colour of the host window and use it for the input EditText
        TypedValue typedValue = new TypedValue();
        activity.getTheme().resolveAttribute(android.R.attr.windowBackground, typedValue, true);
        if (typedValue.type >= TypedValue.TYPE_FIRST_COLOR_INT
                && typedValue.type <= TypedValue.TYPE_LAST_COLOR_INT) {
            // windowBackground is a color
            mPickerEdit.setBackgroundColor(typedValue.data);
        } else {
            // windowBackground is not a color, probably a drawable
            mPickerEdit.setBackground(ContextCompat.getDrawable(activity, typedValue.resourceId));
        }

        if(attrs != null) {
            TypedArray a =
                    getContext().obtainStyledAttributes(attrs, R.styleable.WeightPicker, 0, 0);
            mPrecision = a.getInt(R.styleable.WeightPicker_precision, 1);
            mBase = a.getInt(R.styleable.WeightPicker_modDivisor, 1);

            setMinValue(a.getInt(R.styleable.WeightPicker_minValue, MIN_MIN_VALUE));
            setMaxValue(a.getInt(R.styleable.WeightPicker_maxValue, MAX_MAX_VALUE));
            mInitialSetup = false;

            setDisplayValues(mHectograms, mBase);

            setValue(a.getInt(R.styleable.WeightPicker_initialValue, mMinValue));
            mKilograms.setWrapSelectorWheel(false);

            a.recycle();
        }

        mHectograms.setOnValueChangedListener(new NumberPicker.OnValueChangeListener() {
            @Override
            public void onValueChange(NumberPicker numberPicker, int oldValue, int newValue) {
                mKilograms.setValue(getWholeKilograms(newValue));
                if (mListener != null) {
                    mListener.onValueChange(WeightPicker.this,
                                            oldValue * mPrecision * mBase,
                                            newValue * mPrecision * mBase);
                }
            }
        });

        mKilograms.setOnValueChangedListener(new NumberPicker.OnValueChangeListener() {
            @Override
            public void onValueChange(NumberPicker numberPicker, int i0, int i1) {
                int oldValue = WeightPicker.this.getValue();
                mHectograms.setValue(mHectograms.getValue() + (i1 - i0) * mBase );
                if (mListener != null) {
                    mListener.onValueChange(WeightPicker.this,
                                            oldValue,
                                            WeightPicker.this.getValue());
                }
            }
        });
    }

    public void setMinValue(int value) {
        mMinValue = value <  MIN_MIN_VALUE ?  MIN_MIN_VALUE : value;
        mHectograms.setMinValue(toPickerValue(mMinValue, mPrecision));
        mKilograms.setMinValue(getWholeKilograms(mHectograms.getMinValue()));
        if(!mInitialSetup) {
            setDisplayValues(mHectograms, mBase);
        }
    }

    private int toPickerValue(int actualValue, int precision) {
        return (int) Math.round((double) actualValue / precision);
    }

    private int getWholeKilograms(int hectograms) {
        return (hectograms - hectograms % mBase) / mBase;
    }

    public void setMaxValue(int value) {
        mMaxValue = value > mMinValue ? value : mMinValue;
        mHectograms.setMaxValue(toPickerValue(mMaxValue, mPrecision));
        mKilograms.setMaxValue(getWholeKilograms(mHectograms.getMaxValue()));
        if(!mInitialSetup) {
            setDisplayValues(mHectograms, mBase);
        }
    }

    private void setDisplayValues(NumberPicker picker, int base) {
        int minVal = picker.getMinValue();
        int maxVal = picker.getMaxValue();

        int length = maxVal - minVal + 1;

        if (length <= 0) {
            return;
        }

        String displayValuesArray[] = new String[length];

        for (int i = 0; i < displayValuesArray.length; ++i) {
            displayValuesArray[i] = String.valueOf((picker.getMinValue() + i) % base);
        }
        picker.setDisplayedValues(displayValuesArray);
        picker.setWrapSelectorWheel(false);
    }

    public int getMinValue() {
        return mMinValue;
    }

    public void setValue (int value) {
        if (value < mMinValue) {
            value = mMinValue;
        } else if (value > mMaxValue) {
            value = mMaxValue;
        }
        mHectograms.setValue(toPickerValue(value, mPrecision));
        mKilograms.setValue(getWholeKilograms(mHectograms.getValue()));
    }

    public int getValue() {
        return mHectograms.getValue() * mPrecision;
    }

    @Override
    protected Parcelable onSaveInstanceState() {
        int value = getValue();
        Bundle state = new Bundle();

        state.putParcelable(SUPERSTATE, super.onSaveInstanceState());
        state.putInt(VALUE, value);

        return (state);
    }

    @Override
    protected void onRestoreInstanceState(Parcelable st) {
        Bundle state = (Bundle) st;
        super.onRestoreInstanceState(state.getParcelable(SUPERSTATE));
        setValue(state.getInt(VALUE));
    }

    public OnValueChangedListener getOnValueChangedListener() {
        return mListener;
    }


    public void setOnValueChangedListener (OnValueChangedListener listener) {
        mListener = listener;
    }

    /**
     * The callback interface used to indicate the value has been changed
     */
    public interface OnValueChangedListener {
        void onValueChange(WeightPicker picker, int oldValue, int newValue);
    }

    private EditText findInput (ViewGroup np) {
        int count = np.getChildCount();
        for (int i = 0; i < count; i++) {
            final View child = np.getChildAt(i);
            if (child instanceof ViewGroup) {
                findInput((ViewGroup) child);
            } else if (child instanceof EditText) {
                return (EditText) child;
            }
        }
        return null;
    }
}

File: weightpicker/src/main/java/co/gbyte/weightpicker/WeightPicker.kt
package co.gbyte.weightpicker

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.os.Parcelable
import android.util.AttributeSet
import android.util.TypedValue
import android.view.View
import android.view.ViewGroup
import android.widget.NumberPicker
import android.widget.RelativeLayout
import android.widget.TextView
import java.text.DecimalFormat
import java.text.NumberFormat
import java.util.*
import android.widget.EditText
import android.view.inputmethod.EditorInfo
import android.view.WindowManager
import android.view.inputmethod.InputMethodManager
import androidx.core.content.ContextCompat

class WeightPicker constructor(context: Context, attrs: AttributeSet?, defStyle: Int)
        : RelativeLayout(context, attrs, defStyle) {

    companion object {
        private const val SUPERSTATE = "superState"
        private const val VALUE = "value"
        private const val MIN_MIN_VALUE = 0
        private const val MAX_MAX_VALUE = 999949
    }

    var value: Int = 0
        set(value) {
            field = when {
                value < _minValue -> _minValue
                value > _maxValue -> _maxValue
                else -> value
            }
            _hectograms.value = toPickerValue(field)
            _kilograms.value = getWholeKilograms(value)
        }

    private fun toPickerValue(actualValue: Int): Int {
        return Math.round(actualValue.toDouble() / _precision ).toInt()
    }

    private fun getWholeKilograms (grams: Int): Int{
        return (grams - grams % (_base * _precision)) / (_base * _precision)
    }

    private var _minValue = MIN_MIN_VALUE
        set(value) {
            field = when {
                value < MIN_MIN_VALUE -> MIN_MIN_VALUE
                value > _maxValue -> _maxValue
                value > MAX_MAX_VALUE -> MAX_MAX_VALUE
                else -> value
            }
            _hectograms.minValue = toPickerValue(field)
            _kilograms.minValue = getWholeKilograms(field)
        }

    private var _maxValue = MAX_MAX_VALUE
        set(value) {
            field = when {
                value > MAX_MAX_VALUE -> MAX_MAX_VALUE
                value < _minValue -> _minValue
                value < MIN_MIN_VALUE -> MIN_MIN_VALUE
                else -> value
            }
            _hectograms.maxValue = toPickerValue(field)
            _kilograms.maxValue = getWholeKilograms(field)
            if(!_initialSetup) setDisplayValues(_hectograms, _base)
        }

    lateinit var onValueChangedListener: OnValueChangedListener

    private var _precision = 1
    private var _base = 1
    private var _oldValue = _minValue
    private val _hectograms: NumberPicker
    private val _kilograms: NumberPicker
    private val _pickerEdit: EditText
    private var _initialSetup = true

    constructor (context: Context, attrs: AttributeSet?) : this(context, attrs, 0)
    constructor (context: Context) : this(context, null)
    init {
        val activity = context as Activity
        activity.layoutInflater.inflate(R.layout.weightpicker, this, true)

        val separator: TextView = findViewById(R.id.separator)
        val numberFormat = NumberFormat.getInstance(Locale.getDefault())
        if (numberFormat is DecimalFormat) {
            val sep = numberFormat.decimalFormatSymbols.decimalSeparator
            separator.text = String.format("%c", sep)
        }

        _kilograms = findViewById(R.id.kilograms)
        _hectograms = findViewById(R.id.hectograms)
        _pickerEdit = findViewById(R.id.weightPickerEdit)

        _pickerEdit.setOnEditorActionListener { _, i, _ ->
            if (i == EditorInfo.IME_ACTION_DONE) {
                _pickerEdit.visibility = View.GONE
            }
            _pickerEdit.clearFocus()
            true
        }

        _pickerEdit.setSelectAllOnFocus(true)

        _pickerEdit.onFocusChangeListener = OnFocusChangeListener { _, hasFocus ->
            if (hasFocus) {
                _oldValue = this@WeightPicker.value
                (activity.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager)
                        .toggleSoftInput(InputMethodManager.SHOW_FORCED, 0)
                _pickerEdit.selectAll()
            } else {
                activity.window.setSoftInputMode(WindowManager
                        .LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN)
                val value = java.lang.Double.parseDouble(_pickerEdit.text.toString())
                this@WeightPicker.value = (value * _precision * _base).toInt()

                onValueChangedListener
                        .onValueChange(this@WeightPicker, _oldValue, this@WeightPicker.value)

                _pickerEdit.visibility = View.GONE
            }
        }

        _pickerEdit.setSelectAllOnFocus(true)

        val longClickListener = OnLongClickListener {
            _pickerEdit.visibility = View.VISIBLE
            _pickerEdit.setText(
                    String.format(Locale.getDefault(),
                    "%.1f",
                    value.toDouble() / (_precision * _base)))
            _pickerEdit.requestFocus()
            true
        }

        _hectograms.setOnLongClickListener(longClickListener)
        separator.setOnLongClickListener(longClickListener)
        _kilograms.setOnLongClickListener(longClickListener)

        val hectogramsEditText = findInput(_hectograms)

        if (hectogramsEditText != null) {
            hectogramsEditText.isFocusable = false
        }

        val kilogramsEditText = findInput(_kilograms)
        if (kilogramsEditText != null) {
            kilogramsEditText.isFocusable = false
        }

        // Find out the background colour of the host window and use it for the input EditText
        val typedValue = TypedValue()
        activity.theme.resolveAttribute(android.R.attr.windowBackground, typedValue, true)
        if (typedValue.type >= TypedValue.TYPE_FIRST_COLOR_INT
                && typedValue.type <= TypedValue.TYPE_LAST_COLOR_INT) {
            // windowBackground is a color
            _pickerEdit.setBackgroundColor(typedValue.data)
        } else {
            // windowBackground is not a color, probably a drawable
            _pickerEdit.background = ContextCompat.getDrawable(activity, typedValue.resourceId)
        }

        if (attrs != null) {
            val a = getContext().obtainStyledAttributes(attrs, R.styleable.WeightPicker, 0, 0)
            _precision = a.getInt(R.styleable.WeightPicker_precision, 1)
            _base = a.getInt(R.styleable.WeightPicker_modDivisor, 1)

            _minValue = a.getInt(R.styleable.WeightPicker_minValue, MIN_MIN_VALUE)
            _maxValue = a.getInt(R.styleable.WeightPicker_maxValue, MAX_MAX_VALUE)
            _initialSetup = false

            setDisplayValues(_hectograms, _base)

            value = a.getInt(R.styleable.WeightPicker_initialValue, _minValue)
            _kilograms.wrapSelectorWheel = false

            a.recycle()
        }

        _hectograms.setOnValueChangedListener { _, _, i1 ->
            _kilograms.value = getWholeKilograms(i1 * _precision)
            val oldValue = value
            value = i1 * _precision
            onValueChangedListener.onValueChange(this@WeightPicker, oldValue, this@WeightPicker.value)
        }

        _kilograms.setOnValueChangedListener { _, i0, i1 ->
            val oldValue = value
            value += (i1 - i0) * 1000
            _hectograms.value = toPickerValue(value)
            onValueChangedListener.onValueChange(this@WeightPicker, oldValue, this@WeightPicker.value)
        }
    }

    private fun findInput(np: ViewGroup): EditText? {
        val count = np.childCount
        for (i in 0 until count) {
            val child = np.getChildAt(i)
            if (child is ViewGroup) {
                findInput(child)
            } else if (child is EditText) {
                return child
            }
        }
        return null
    }

    private fun setDisplayValues(picker: NumberPicker, base: Int) {
        val minVal = picker.minValue
        val maxVal = picker.maxValue
        val length = maxVal - minVal + 1

        if (length <= 0) {
            return
        }

        val displayValuesArray = arrayOfNulls<String>(length)

        for (i in displayValuesArray.indices) {
            displayValuesArray[i] = ((picker.minValue + i) % base).toString()
        }
        picker.displayedValues = displayValuesArray
        picker.wrapSelectorWheel = false
    }

    override fun onSaveInstanceState(): Parcelable? {
        val value = value
        val state = Bundle()

        state.putParcelable(SUPERSTATE, super.onSaveInstanceState())
        state.putInt(VALUE, value)

        return state
    }

    override fun onRestoreInstanceState(st: Parcelable) {
        val state = st as Bundle
        super.onRestoreInstanceState(state.getParcelable(SUPERSTATE))
        value = state.getInt(VALUE)
    }

    /**
     * The callback interface used to indicate the value has been changed
     */
    interface OnValueChangedListener {
        fun onValueChange(picker: WeightPicker, oldValue: Int, newValue: Int)
    }
}
